composer/core/surgery.py

```diff
@@ -142,3 +188,159 @@ def count_module_instances(model: torch.nn.Module, module_class: Type[torch.nn.M
         count += count_module_instances(child, module_class)
 
     return count
+
+
+def _tensor_in(tensor: torch.Tensor, iterable: Iterable[torch.Tensor]):
+    """Returns whether `tensor is element` for any element in `iterable`
+    This function is necessary because `tensor in iterable` does not work
+    reliably for `Tensor`s.
+    See https://discuss.pytorch.org/t/how-to-judge-a-tensor-is-in-a-list/15998/4
+    for further discussion.
+    """
+    return any(tensor is elem for elem in iterable)
+
+
+def _find_param_in_optimizer(param: torch.nn.parameter.Parameter, optimizer: torch.optim.Optimizer) -> int:
+    """Returns the index of the optimizer ``param_group`` containing ``param``
+    Optimizers store their parameters within an iterable of ``dict``s called
+    :attr:`~torch.optim.Optimizer.param_groups`.
+    By default, there is only one group in :attr:`~torch.optim.Optimizer.param_groups`
+    that containing all the parameters, but there can be more than one. This
+    function is a simple utility to identify which parameter group in
+    :attr:`~torch.optim.Optimizer.param_groups` contains a given parameter, if any. The information
+    might be desirable to, e.g., inspect the optimizer settings being used
+    for a given parameter, or to remove unused parameter tensors from
+    the optimizer.
+
+    Args:
+        param (torch.nn.parameter.Parameter): The parameter to search for.
+        optimizer (torch.optim.Optimizer): The optimizer to search within.
+    
+    Returns:
+        int: The index within `opt.param_groups` of the first group containing ``param``,
+        or `-1` if ``param`` is not in the ``opt`.
+    """
+    for i, group in enumerate(optimizer.param_groups):
+        param_list: List[torch.nn.parameter.Parameter] = group['params']
+        if _tensor_in(param, param_list):
+            return i
+
+    return -1
+
+
+def update_params_in_optimizer(old_params: Iterable[torch.nn.parameter.Parameter],
```

Comment:
Thoughts about replacing this function with something that takes the output of `replace_module_classes`? I lm thinking then we would have a new function for explicitley adding parameters E.g. something like:

```python3
def update_params_in_optimizer(old_params_to_replaced_params: Dict[nn.Parameters, Optional[nn.Parameters]])
  ...

def add_paramaters_to_optimizer(param_group_of: nn.Parameter ,new_parameters: Iterable[parameters])
  ...
```

The first function could handle replqcements / removals (if the value of a dict is none, thrn that would act as a removal) across parameters groups, and the second one could add in new parameters. For the 2nd function, you would pass in a parameters that is used to infer the parameter group.

Reply:
I instinctively like the idea of having something directly take in the output of `replace_module_classes`. I'm not sure when we'd call it though.

I'm mildly skeptical of having a public function to support adding parameters to a specific param group, as opposed to only adding them to "the same group as these other params I'm giving you". Mostly because:

1. I don't think the caller should have to keep track of param groups, especially since we don't return them (and I wouldn't be surprised if many users didn't even know about param_groups).
2. Conditioned on the user not managing param_groups, they should probably be calling `add_param_group` when adding params to stay within the optimizer's public interface.

Brainstorming a little bit, I think (?) we have two use cases:

1. Making module replacement not ruin the set of tensors being optimized
2. Making moving params from one device to another not ruin the set of tensors being optimized

(2) could be handled in the same manner as (1), but it might be good to just replace the groups wholesale to depend a bit less on implementation details. If there are other use cases, that would definitely suggest building more general functionality.

So...I don't actually know the ideal solution, but that's my 2c.

Reply:
Agreed with @dblalock that a public function to `add_param_group` is duplicative of the optimizer's public interface. Going to close this API discussion, can file an issue if someone feels strongly here.